rep/event.go

```diff
@@ -18,35 +16,151 @@ const (
+	return RetrievalEventConnect{phase, payloadCid, storageProviderId, storageProviderAddr}
+}
+
+type RetrievalEventQueryAsk struct {
+	phase               Phase
+	payloadCid          cid.Cid
+	storageProviderId   peer.ID
+	storageProviderAddr address.Address
+	queryResponse       retrievalmarket.QueryResponse
+}
+
+func NewRetrievalEventQueryAsk(phase Phase, payloadCid cid.Cid, storageProviderId peer.ID, storageProviderAddr address.Address, queryResponse retrievalmarket.QueryResponse) RetrievalEventQueryAsk {
+	return RetrievalEventQueryAsk{phase, payloadCid, storageProviderId, storageProviderAddr, queryResponse}
+}
+
+type RetrievalEventProposed struct {
+	phase               Phase
+	payloadCid          cid.Cid
+	storageProviderId   peer.ID
+	storageProviderAddr address.Address
+}
+
+func NewRetrievalEventProposed(phase Phase, payloadCid cid.Cid, storageProviderId peer.ID, storageProviderAddr address.Address) RetrievalEventProposed {
+	return RetrievalEventProposed{phase, payloadCid, storageProviderId, storageProviderAddr}
+}
+
+type RetrievalEventAccepted struct {
+	phase               Phase
+	payloadCid          cid.Cid
+	storageProviderId   peer.ID
+	storageProviderAddr address.Address
+}
+
+func NewRetrievalEventAccepted(phase Phase, payloadCid cid.Cid, storageProviderId peer.ID, storageProviderAddr address.Address) RetrievalEventAccepted {
+	return RetrievalEventAccepted{phase, payloadCid, storageProviderId, storageProviderAddr}
+}
+
+type RetrievalEventFirstByte struct {
+	phase               Phase
+	payloadCid          cid.Cid
+	storageProviderId   peer.ID
+	storageProviderAddr address.Address
+}
+
+func NewRetrievalEventFirstByte(phase Phase, payloadCid cid.Cid, storageProviderId peer.ID, storageProviderAddr address.Address) RetrievalEventFirstByte {
+	return RetrievalEventFirstByte{phase, payloadCid, storageProviderId, storageProviderAddr}
+}
+
+type RetrievalEventFailure struct {
+	phase               Phase
+	payloadCid          cid.Cid
+	storageProviderId   peer.ID
+	storageProviderAddr address.Address
+	errorMessage        string
+}
+
+func NewRetrievalEventFailure(phase Phase, payloadCid cid.Cid, storageProviderId peer.ID, storageProviderAddr address.Address, errorMessage string) RetrievalEventFailure {
+	return RetrievalEventFailure{phase, payloadCid, storageProviderId, storageProviderAddr, errorMessage}
+}
+
+type RetrievalEventSuccess struct {
+	phase               Phase
+	payloadCid          cid.Cid
+	storageProviderId   peer.ID
+	storageProviderAddr address.Address
+	receivedSize        uint64
+}
+
+func NewRetrievalEventSuccess(phase Phase, payloadCid cid.Cid, storageProviderId peer.ID, storageProviderAddr address.Address, receivedSize uint64) RetrievalEventSuccess {
+	return RetrievalEventSuccess{phase, payloadCid, storageProviderId, storageProviderAddr, receivedSize}
 }
+
+func (r RetrievalEventConnect) Code() Code                                    { return ConnectCode }
```

Comment:
I wonder if we need to go as far as having all these interface methods instead of just public data fields. My baseline feeling is for simple data objects like events, I'm leaning towards public vars and not methods (among other things, serializes well if you need to do some reflect/bindnode madness). It's fine honestly though either way. 

In terms of constraining the type, I'm starting to lean on having a single private method interface method -- it's a total hack, but the closest I've gotten with go to a simple method of algebraic data types.

Anyway, just food for thought.

Reply:
So, I actually started like that, all of the events had the same public properties - `Phase`, `StorageProviderID`, `PayloadCid`, although I left `Code()` as a method to give the interface something to describe, and it was working fine, until I had to consume the events in the event listener. See `OnRetrievalEvent()`. To get access to any of the properties I had to implement a type switch with a case for each variation and then end up repeating the ~4 main properties in each case because I couldn't get access to them generically! So I flipped over to this version where there's generic access via interface methods and the type switch only needs to happen to get access to the special properties on 3 of the events.

If interfaces did public properties, not just methods, that'd be awesome, but it seems like this is a pain point, and either way you have to decide whether your internal implementation has _tons_ of code, or your downstream user needs to have _tons_ of code. So I chose the former but :shrug:. If you have better ideas then I'm all ears.

Reply:
fair.